File: backend/src/main/java/com/hotelmanagement/model/Allocation.java
package com.hotelmanagement.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Allocation Entity - Represents a hostel room allocation for students
 * Table: allocations
 */
@Entity
@Table(name = "allocations")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Allocation {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "room_id", nullable = false)
    private Room room;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;
    
    @Column(name = "allocation_date", nullable = false)
    private LocalDate allocationDate;
    
    @Column(name = "end_date", nullable = false)
    private LocalDate endDate;
    
    @Column(nullable = false)
    private String status = "ACTIVE"; // ACTIVE, VACATED, PENDING
    
    @Column(name = "academic_year", nullable = false, length = 20)
    private String academicYear; // 2024-2025
    
    @Column(name = "semester", nullable = false, length = 20)
    private String semester; // ODD, EVEN
    
    @Column(name = "special_requests", columnDefinition = "TEXT")
    private String specialRequests;
    
    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;
    
    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: node-backend/src/models/Allocation.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');
const Room = require('./Room');
const Student = require('./Student');

const Allocation = sequelize.define('Allocation', {
  id: {
    type: DataTypes.BIGINT,
    primaryKey: true,
    autoIncrement: true
  },
  roomId: {
    type: DataTypes.BIGINT,
    allowNull: false,
    field: 'room_id',
    references: {
      model: Room,
      key: 'id'
    }
  },
  studentId: {
    type: DataTypes.BIGINT,
    allowNull: false,
    field: 'student_id',
    references: {
      model: Student,
      key: 'id'
    }
  },
  allocationDate: {
    type: DataTypes.DATEONLY,
    allowNull: false,
    field: 'allocation_date'
  },
  endDate: {
    type: DataTypes.DATEONLY,
    allowNull: false,
    field: 'end_date'
  },
  status: {
    type: DataTypes.STRING(50),
    allowNull: false,
    defaultValue: 'ACTIVE'
  },
  academicYear: {
    type: DataTypes.STRING(20),
    allowNull: false,
    field: 'academic_year'
  },
  semester: {
    type: DataTypes.STRING(20),
    allowNull: false
  },
  specialRequests: {
    type: DataTypes.TEXT,
    allowNull: true,
    field: 'special_requests'
  }
}, {
  tableName: 'allocations',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: 'updated_at'
});

// Define associations
Allocation.belongsTo(Room, { foreignKey: 'roomId', as: 'room' });
Allocation.belongsTo(Student, { foreignKey: 'studentId', as: 'student' });
Room.hasMany(Allocation, { foreignKey: 'roomId', as: 'allocations' });
Student.hasMany(Allocation, { foreignKey: 'studentId', as: 'allocations' });

module.exports = Allocation;
